File: Libs/ADLXWrapperCpp/ADLXWrapper.cpp
#include "ADLXWrapper.hpp"

using namespace adlx;

void Initialize()
{
	// When OnInit is called on DLL_PROCESS_ATTACH in dllmain
	// g_ADLXHelp.Initialize() takes 10 seconds.

	OnInit();
}

int GetId(const CHAR* pnpString)
{
	for (auto i = g_gpuList->Begin(); i != g_gpuList->End(); ++i)
	{
		ADLX_RESULT res = ADLX_FAIL;
		IADLXGPUPtr gpu;
		res = g_gpuList->At(i, &gpu);
		if (ADLX_FAILED(res) || gpu == nullptr)
			continue;

		const char* pnp = nullptr;
		res = gpu->PNPString(&pnp);
		if (ADLX_SUCCEEDED(res) && strcmp(pnp, pnpString) == 0)
		{
			int id;
			res = gpu->UniqueId(&id);
			if (ADLX_SUCCEEDED(res))
				return id;
		}
	}
	return -1;
}

void GetGPUs(int* result[], int* count)
{
	*count = (int)g_gpu_ids.size();

	auto size = *count * sizeof(int);
	*result = static_cast<int*>(CoTaskMemAlloc(size));
	if (*result != nullptr)
	{
		memcpy(*result, g_gpu_ids.data(), size);
	}
}

BOOL IsSupported(int id)
{
	auto gpu = FindGPU(id);
	if (gpu == nullptr)
		return FALSE;
	return (BOOL)gpu->IsSupported();
}

BOOL GetFanRange(int id, FanRange* result)
{
	auto gpu = FindGPU(id);
	if (gpu == nullptr)
		return FALSE;

	auto fanSpeedRange = gpu->GetFanSpeedRange();
	auto fanTemperatureRange = gpu->GetFanTemperatureRange();
	auto newResult = FanRange();
	newResult.minTemperature = fanTemperatureRange.minValue;
	newResult.maxTemperature = fanTemperatureRange.maxValue;
	newResult.minSpeed = fanSpeedRange.minValue;
	newResult.maxSpeed = fanSpeedRange.maxValue;

	memcpy(result, &newResult, sizeof(newResult));
	return TRUE;
}

BOOL GetFanSpeeds(int id, FanSpeeds* result)
{
	auto gpu = FindGPU(id);
	if (gpu == nullptr)
		return FALSE;

	auto s = gpu->GetFanSpeeds();
	auto speeds = FanSpeeds();
	speeds.P1 = s.at(0);
	speeds.P2 = s.at(1);
	speeds.P3 = s.at(2);
	speeds.P4 = s.at(3);
	speeds.P5 = s.at(4);

	memcpy(result, &speeds, sizeof(speeds));
	return TRUE;
}

void SetFanSpeeds(int id, FanSpeeds* speeds)
{
	auto gpu = FindGPU(id);
	if (gpu == nullptr) return;

	std::vector<std::pair<int, int>> items
	{
		std::pair<int, int>(speeds->P1.first, speeds->P1.second),
		std::pair<int, int>(speeds->P2.first, speeds->P2.second),
		std::pair<int, int>(speeds->P3.first, speeds->P3.second),
		std::pair<int, int>(speeds->P4.first, speeds->P4.second),
		std::pair<int, int>(speeds->P5.first, speeds->P5.second)
	};
	gpu->SetFanSpeeds(items);
}

void SetFanSpeed(int id, int speed)
{
	auto gpu = FindGPU(id);
	if (gpu == nullptr) return;

	gpu->SetFanSpeed(speed);
}

BOOL IsZeroRPMSupported(int id)
{
	auto gpu = FindGPU(id);
	if (gpu == nullptr)
		return FALSE;

	return gpu->IsZeroRPMSupported() ? TRUE : FALSE;
}

BOOL IsZeroRPMEnabled(int id)
{
	auto gpu = FindGPU(id);
	if (gpu == nullptr)
		return FALSE;

	return gpu->IsZeroRPMEnabled() ? TRUE : FALSE;
}

void SetZeroRPM(int id, BOOL enable)
{
	auto gpu = FindGPU(id);
	if (gpu == nullptr) return;

	gpu->SetZeroRPM(enable == 0 ? false : true);
}

void OnInit()
{
	ADLX_RESULT res = ADLX_FAIL;
	res = g_ADLXHelp.Initialize();
	if (ADLX_FAILED(res))
		throw("g_ADLXHelp initialize failed");

	res = g_ADLXHelp.GetSystemServices()->GetGPUTuningServices(&g_gpuTuningService);
	if (ADLX_FAILED(res))
		throw("GetGPUTuningServices failed");

	res = g_ADLXHelp.GetSystemServices()->GetGPUs(&g_gpuList);
	if (ADLX_FAILED(res))
		throw("GetGPUs failed");

	for (auto i = g_gpuList->Begin(); i != g_gpuList->End(); ++i)
	{
		IADLXGPUPtr gpu;
		res = g_gpuList->At(i, &gpu);
		if (ADLX_FAILED(res) || gpu == nullptr)
			continue;

		int uniqueId;
		res = gpu->UniqueId(&uniqueId);

		g_gpu_map.emplace(uniqueId, new AMDGPU(g_gpuTuningService, gpu));
		g_gpu_ids.push_back(uniqueId);
	}
}

void OnExit()
{
	for (auto& pair : g_gpu_map)
	{
		AMDGPU* gpu = pair.second;
		delete gpu;
	}
	g_gpuList->Release();
	g_gpuTuningService->Release();
	g_ADLXHelp.Terminate();
}


AMDGPU* FindGPU(int id)
{
	auto it = g_gpu_map.find(id);
	if (it == g_gpu_map.end() || !it->second->IsSupported())
		return nullptr;
	return it->second;
}

File: Libs/ADLXWrapper/ADLXWrapper.cs
using System.Runtime.InteropServices;

namespace ADLXWrapper;

public static partial class ADLXWrapper
{
    [StructLayout(LayoutKind.Sequential)]
    public struct FanRange
    {
        public int minTemperature;
        public int maxTemperature;
        public int minSpeed;
        public int maxSpeed;
    }

    [StructLayout(LayoutKind.Sequential)]
    public struct FanSpeeds
    {
        public KeyValuePair<int, int> P1;
        public KeyValuePair<int, int> P2;
        public KeyValuePair<int, int> P3;
        public KeyValuePair<int, int> P4;
        public KeyValuePair<int, int> P5;
    }

    #region Public methods

    public static void Initialize() => Initialize_();

    public static int GetId(string pnpString) => GetId_(pnpString);

    public static List<int> GetGPUs()
    {
        GetGPUs_(out var items, out var _);
        return items.ToList();
    }

    public static bool IsSupported(int id) => IsSupported_(id);

    public static FanRange? GetFanRange(int id)
    {
        var pRange = Marshal.AllocCoTaskMem(Marshal.SizeOf(typeof(FanRange)));
        if (!GetFanRange_(id, pRange))
        {
            return null;
        }

        return (FanRange?)Marshal.PtrToStructure(pRange, typeof(FanRange));
    }

    public static FanSpeeds? GetFanSpeeds(int id)
    {
        var pSpeeds = Marshal.AllocCoTaskMem(Marshal.SizeOf(typeof(FanSpeeds)));
        if (!GetFanSpeeds_(id, pSpeeds))
        {
            return null;
        }

        return (FanSpeeds?)Marshal.PtrToStructure(pSpeeds, typeof(FanSpeeds));
    }

    public static void SetFanSpeeds(int id, FanSpeeds speeds)
    {
        var pSpeeds = Marshal.AllocCoTaskMem(Marshal.SizeOf(typeof(FanSpeeds)));
        Marshal.StructureToPtr(speeds, pSpeeds, false);
        SetFanSpeeds_(id, pSpeeds);
    }

    public static void SetFanSpeed(int id, int speed) => SetFanSpeed_(id, speed);

    public static bool IsZeroRPMSupported(int id) => IsZeroRPMSupported_(id);

    public static bool IsZeroRPMEnabled(int id) => IsZeroRPMEnabled_(id);

    public static void SetZeroRPM(int id, bool enable) => SetZeroRPM_(id, enable);

    #endregion

    #region P/Invoke methods

    [LibraryImport("ADLXWrapperCpp.dll", EntryPoint = "Initialize")]
    private static partial void Initialize_();

    [LibraryImport("ADLXWrapperCpp.dll", EntryPoint = "GetId")]
    private static partial int GetId_([MarshalAs(UnmanagedType.LPStr)] string pnpString);

    [LibraryImport("ADLXWrapperCpp.dll", EntryPoint = "GetGPUs")]
    private static partial void GetGPUs_(
        [MarshalAs(UnmanagedType.LPArray, SizeParamIndex = 1)] out int[] result,
        out int count);

    [LibraryImport("ADLXWrapperCpp.dll", EntryPoint = "IsSupported")]
    [return: MarshalAs(UnmanagedType.Bool)]
    private static partial bool IsSupported_(int id);

    [LibraryImport("ADLXWrapperCpp.dll", EntryPoint = "GetFanRange")]
    [return: MarshalAs(UnmanagedType.Bool)]
    private static partial bool GetFanRange_(int id, IntPtr result);

    [LibraryImport("ADLXWrapperCpp.dll", EntryPoint = "GetFanSpeeds")]
    [return: MarshalAs(UnmanagedType.Bool)]
    private static partial bool GetFanSpeeds_(int id, IntPtr result);

    [LibraryImport("ADLXWrapperCpp.dll", EntryPoint = "SetFanSpeeds")]
    private static partial void SetFanSpeeds_(int id, IntPtr speeds);

    [LibraryImport("ADLXWrapperCpp.dll", EntryPoint = "SetFanSpeed")]
    private static partial void SetFanSpeed_(int id, int speed);

    [LibraryImport("ADLXWrapperCpp.dll", EntryPoint = "IsZeroRPMSupported")]
    [return: MarshalAs(UnmanagedType.Bool)]
    private static partial bool IsZeroRPMSupported_(int id);

    [LibraryImport("ADLXWrapperCpp.dll", EntryPoint = "IsZeroRPMEnabled")]
    [return: MarshalAs(UnmanagedType.Bool)]
    private static partial bool IsZeroRPMEnabled_(int id);

    [LibraryImport("ADLXWrapperCpp.dll", EntryPoint = "SetZeroRPM")]
    private static partial void SetZeroRPM_(
        int id,
        [MarshalAs(UnmanagedType.Bool)] bool enable);

    #endregion
}
